codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -170,6 +170,7 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                                 "Stream %d in unexpected state: %s", stream.id(), stream.state()));
                 }
             }
+            stream.headersSent();
```

Comment:
Your understanding is generally correct. However enqueueing headers in the flow controller is primarily to preserve the ordering of trailers relative to data, and for the initial headers we can typically avoid going through the flow controller. Either way if we enqueue or "directly write" it is still not necessarily an "immediate" operation (depends on networking conditions, flush calls, etc...). The motivation for delaying the call to `headersSent()` is if the call to write (or queue a write in flow control) fail for some reason (even if its unlikely), the peer will not receive headers, and so headers have not actually been sent, and we shouldn't necessarily prevent future attempts to send headers if these conditions are corrected. Either way the `headersSent` is a synchronous operation (marked immediately) and the write call is asynchronous.

We do the same thing with [resetSent](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java#L654).